crates/resolver/src/system_conf/android.rs

```diff
@@ -0,0 +1,91 @@
+use std::net::IpAddr;
+use std::sync::Arc;
+use tracing::trace;
+
+use crate::config::{NameServerConfig, ResolverConfig, ResolverOpts};
+use crate::proto::ProtoError;
+
+pub fn read_system_conf() -> Result<(ResolverConfig, ResolverOpts), ProtoError> {
+    read_android_conf().map_err(|e| ProtoError::from(Arc::new(e)))
+}
+
+fn read_android_conf() -> Result<(ResolverConfig, ResolverOpts), jni::errors::Error> {
+    let ctx = ndk_context::android_context();
+    let vm = unsafe { jni::JavaVM::from_raw(ctx.vm().cast()) }?;
+    let mut env = vm.attach_current_thread()?;
+
+    let activity = unsafe { JObject::from_raw(ctx.context().cast()) };
+
+    // https://developer.android.com/reference/android/content/Context#getSystemService(java.lang.String)
+    let connectivity_service = env.new_string("connectivity")?;
+    let connectivity_manager = env
+        .call_method(
+            activity,
+            "getSystemService",
+            "(Ljava/lang/String;)Ljava/lang/Object;",
+            &[JValue::Object(&connectivity_service)],
+        )?
+        .l()?;
+
+    // https://developer.android.com/reference/android/net/ConnectivityManager#getActiveNetwork()
+    let network = env
+        .call_method(
+            &connectivity_manager,
+            "getActiveNetwork",
+            "()Landroid/net/Network;",
+            &[],
+        )?
+        .l()?;
+
+    // https://developer.android.com/reference/android/net/ConnectivityManager#getLinkProperties(android.net.Network)
+    let link_properties = env
+        .call_method(
+            &connectivity_manager,
+            "getLinkProperties",
+            "(Landroid/net/Network;)Landroid/net/LinkProperties;",
+            &[JValue::Object(&network)],
+        )?
+        .l()?;
+
+    // https://developer.android.com/reference/android/net/LinkProperties#getDnsServers()
+    let dns_servers = env
+        .call_method(&link_properties, "getDnsServers", "()Ljava/util/List;", &[])?
+        .l()?;
+    let dns_servers = env.get_list(&dns_servers)?;
+    let mut dns_servers = dns_servers.iter(&mut env)?;
+
+    let mut nameservers = Vec::<NameServerConfig>::new();
+    while let Some(server) = dns_servers.next(&mut env)? {
+        let server: AutoLocal<'_, JObject<'_>> = env.auto_local(server);
+
+        // https://developer.android.com/reference/java/net/InetAddress#getAddress()
+        let ip_bytes_obj = env
+            .call_method(&server, "getAddress", "()[B", &[])?
+            .l()?;
+        let ip_bytes = env.convert_byte_array(JPrimitiveArray::<'_, jbyte>::from(ip_bytes_obj))?;
+
+        let ip = match ip_bytes.len() {
+            4 => {
+                let mut arr = [0u8; 4];
+                arr.copy_from_slice(&ip_bytes);
+                IpAddr::from(arr)
+            }
+            16 => {
+                let mut arr = [0u8; 16];
+                arr.copy_from_slice(&ip_bytes);
+                IpAddr::from(arr)
+            }
+            _ => continue,
```

Comment:
Should we yield an error if the length is not 4 or 16?

Reply:
Unsure. This should not happen unless the system is somehow broken, maybe log a warning to allow resolution to still work in this impossible case?